File: insani-backend/app/routers/discrepancies.py
"""
Discrepancies Router — Spec vs Submittal comparison.

POST   /v1/discrepancies/analyze          — Start a new comparison
GET    /v1/discrepancies?project_id=X     — List reports for a project
GET    /v1/discrepancies/{id}             — Get full report with findings
PATCH  /v1/discrepancies/items/{id}       — Update a finding's status
DELETE /v1/discrepancies/{id}             — Delete a report
"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from pydantic import BaseModel
import structlog

from app.db import get_db
from app.services import discrepancy_service
from app.middleware.auth import require_auth_context, AuthContext

logger = structlog.get_logger()
router = APIRouter(prefix="/v1/discrepancies", tags=["Discrepancies"])


# ── Schemas ──

class AnalyzeRequest(BaseModel):
    project_id: int
    spec_doc_ids: list[int]
    submittal_doc_ids: list[int]
    title: str = ""


class ItemUpdateRequest(BaseModel):
    status: str  # open, acknowledged, resolved, dismissed


class DiscrepancyItemResponse(BaseModel):
    id: int
    severity: str
    category: str
    title: str
    description: str
    spec_reference: str
    spec_doc_id: int | None
    spec_page: int | None
    spec_excerpt: str
    submittal_reference: str
    submittal_doc_id: int | None
    submittal_page: int | None
    submittal_excerpt: str
    recommendation: str
    status: str


class ReportResponse(BaseModel):
    id: int
    title: str
    status: str
    summary: str
    discrepancy_count: int
    spec_doc_ids: list[int]
    submittal_doc_ids: list[int]
    created_at: str | None = None
    items: list[DiscrepancyItemResponse] = []


class ReportListItem(BaseModel):
    id: int
    title: str
    status: str
    discrepancy_count: int
    created_at: str | None = None


# ── Endpoints ──

@router.post("/analyze", response_model=ReportResponse, status_code=201)
async def analyze(
    body: AnalyzeRequest,
    ctx: AuthContext = Depends(require_auth_context),
    db: AsyncSession = Depends(get_db),
):
    """Start a spec vs submittal comparison. Runs synchronously and returns the full report."""
    if not body.spec_doc_ids:
        raise HTTPException(status_code=400, detail="At least one spec document is required")
    if not body.submittal_doc_ids:
        raise HTTPException(status_code=400, detail="At least one submittal document is required")

    # Create report
    report = await discrepancy_service.create_report(
        db, ctx.org_id, body.project_id, ctx.user_id,
        body.spec_doc_ids, body.submittal_doc_ids, body.title,
    )
    await db.flush()

    # Run analysis
    report = await discrepancy_service.run_analysis(db, report.id, ctx.org_id)
    await db.commit()

    # Reload with items
    report = await discrepancy_service.get_report(db, report.id, ctx.org_id)

    return _report_to_response(report)


@router.get("/", response_model=list[ReportListItem])
async def list_reports(
    project_id: int,
    ctx: AuthContext = Depends(require_auth_context),
    db: AsyncSession = Depends(get_db),
):
    """List all discrepancy reports for a project."""
    reports = await discrepancy_service.list_reports(db, project_id, ctx.org_id)
    return [
        ReportListItem(
            id=r.id,
            title=r.title,
            status=r.status,
            discrepancy_count=r.discrepancy_count,
            created_at=str(r.created_at) if r.created_at else None,
        )
        for r in reports
    ]


@router.get("/{report_id}", response_model=ReportResponse)
async def get_report(
    report_id: int,
    ctx: AuthContext = Depends(require_auth_context),
    db: AsyncSession = Depends(get_db),
):
    """Get a full report with all findings."""
    report = await discrepancy_service.get_report(db, report_id, ctx.org_id)
    if not report:
        raise HTTPException(status_code=404, detail="Report not found")
    return _report_to_response(report)


@router.patch("/items/{item_id}")
async def update_item(
    item_id: int,
    body: ItemUpdateRequest,
    ctx: AuthContext = Depends(require_auth_context),
    db: AsyncSession = Depends(get_db),
):
    """Update a finding's status (resolve, dismiss, etc.)."""
    if body.status not in ("open", "acknowledged", "resolved", "dismissed"):
        raise HTTPException(status_code=400, detail="Invalid status")

    item = await discrepancy_service.update_item_status(db, item_id, ctx.org_id, body.status, ctx.user_id)
    if not item:
        raise HTTPException(status_code=404, detail="Item not found")
    await db.commit()

    return {"id": item.id, "status": item.status}


@router.delete("/{report_id}")
async def delete_report(
    report_id: int,
    ctx: AuthContext = Depends(require_auth_context),
    db: AsyncSession = Depends(get_db),
):
    """Delete a report and all its findings."""
    deleted = await discrepancy_service.delete_report(db, report_id, ctx.org_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Report not found")
    await db.commit()
    return {"deleted": True}


def _report_to_response(report) -> ReportResponse:
    items = []
    if hasattr(report, 'items') and report.items:
        items = [
            DiscrepancyItemResponse(
                id=i.id,
                severity=i.severity,
                category=i.category,
                title=i.title,
                description=i.description,
                spec_reference=i.spec_reference,
                spec_doc_id=i.spec_doc_id,
                spec_page=i.spec_page,
                spec_excerpt=i.spec_excerpt,
                submittal_reference=i.submittal_reference,
                submittal_doc_id=i.submittal_doc_id,
                submittal_page=i.submittal_page,
                submittal_excerpt=i.submittal_excerpt,
                recommendation=i.recommendation,
                status=i.status,
            )
            for i in report.items
        ]

    return ReportResponse(
        id=report.id,
        title=report.title,
        status=report.status,
        summary=report.summary,
        discrepancy_count=report.discrepancy_count,
        spec_doc_ids=report.spec_doc_ids or [],
        submittal_doc_ids=report.submittal_doc_ids or [],
        created_at=str(report.created_at) if report.created_at else None,
        items=items,
    )

File: insani-project/js/discrepancies.js
/* ═══════════════════════════════════════════════
   DISCREPANCIES — Spec vs Submittal comparison.
   Upload specs + submittals, run comparison,
   view findings, update statuses.
   ═══════════════════════════════════════════════ */

var discrepancyReports = [];
var activeReport = null;

// ═══ API ═══

async function apiAnalyzeDiscrepancies(projectId, specDocIds, submittalDocIds, title) {
  return apiFetch('/v1/discrepancies/analyze', {
    method: 'POST',
    body: JSON.stringify({
      project_id: projectId,
      spec_doc_ids: specDocIds,
      submittal_doc_ids: submittalDocIds,
      title: title || ''
    })
  });
}

async function apiListDiscrepancyReports(projectId) {
  return apiFetch('/v1/discrepancies?project_id=' + projectId);
}

async function apiGetDiscrepancyReport(reportId) {
  return apiFetch('/v1/discrepancies/' + reportId);
}

async function apiUpdateDiscrepancyItem(itemId, status) {
  return apiFetch('/v1/discrepancies/items/' + itemId, {
    method: 'PATCH',
    body: JSON.stringify({ status: status })
  });
}

async function apiDeleteDiscrepancyReport(reportId) {
  return apiFetch('/v1/discrepancies/' + reportId, { method: 'DELETE' });
}

// ═══ REPORT LIST ═══

async function loadDiscrepancyReports() {
  if (!activeProjectId) return;
  try {
    discrepancyReports = await apiListDiscrepancyReports(activeProjectId);
    renderDiscrepancyList();
  } catch (e) {
    discrepancyReports = [];
    renderDiscrepancyList();
  }
}

function renderDiscrepancyList() {
  var container = document.getElementById('discrepancyList');
  if (!container) return;

  if (!discrepancyReports.length) {
    container.innerHTML = '<div style="color:var(--text-dim);font-size:0.75rem;padding:0.3rem 0.6rem">No comparisons yet</div>';
    return;
  }

  container.innerHTML = discrepancyReports.map(function(r) {
    var badge = r.status === 'complete' ? r.discrepancy_count + ' findings' : r.status;
    var color = r.discrepancy_count > 0 ? 'var(--red)' : 'var(--green)';
    if (r.status !== 'complete') color = 'var(--text-dim)';
    return '<div class="doc-item" onclick="viewDiscrepancyReport(' + r.id + ')">' +
      '<span class="doc-item-icon" style="color:' + color + '">DC</span>' +
      '<span class="doc-item-name">' + esc(r.title) + '</span>' +
      '<span class="doc-item-pages" style="color:' + color + '">' + badge + '</span>' +
      '<button class="sb-delete" onclick="event.stopPropagation();deleteDiscrepancyReport(' + r.id + ')" title="Delete report">×</button>' +
    '</div>';
  }).join('');
}

async function deleteDiscrepancyReport(reportId) {
  try {
    await apiDeleteDiscrepancyReport(reportId);
    await loadDiscrepancyReports();
    if (activeReport && activeReport.id === reportId) {
      activeReport = null;
      showChat();
    }
    showToast('Report deleted');
  } catch (e) {
    showToast('Failed to delete report');
  }
}

// ═══ DOCUMENT SELECTOR MODAL ═══

function showCompareModal() {
  if (!activeProjectId) { showToast('Select a project first'); return; }
  if (!projectDocuments.length) { showToast('Upload documents first'); return; }

  var html = '<div class="disc-modal-content">' +
    '<h3 style="margin:0 0 0.5rem;font-size:1rem;font-weight:500">Compare Spec vs Submittal</h3>' +
    '<p style="color:var(--text-muted);font-size:0.78rem;margin:0 0 1rem">Select which documents are specs and which are submittals</p>' +
    '<div class="disc-doc-grid">' +
    '<div class="disc-col"><div class="disc-col-label">Specifications</div><div id="specCheckboxes">' +
    projectDocuments.map(function(d) {
      return '<label class="disc-check"><input type="checkbox" value="' + d.id + '" name="spec"> ' + esc(d.filename) + '</label>';
    }).join('') +
    '</div></div>' +
    '<div class="disc-col"><div class="disc-col-label">Submittals</div><div id="submittalCheckboxes">' +
    projectDocuments.map(function(d) {
      return '<label class="disc-check"><input type="checkbox" value="' + d.id + '" name="submittal"> ' + esc(d.filename) + '</label>';
    }).join('') +
    '</div></div>' +
    '</div>' +
    '<div style="display:flex;gap:0.5rem;justify-content:flex-end;margin-top:1rem">' +
    '<button class="disc-btn disc-btn-cancel" onclick="closeCompareModal()">Cancel</button>' +
    '<button class="disc-btn disc-btn-run" onclick="runComparison()">Run Comparison</button>' +
    '</div></div>';

  document.getElementById('compareModal').innerHTML = html;
  document.getElementById('compareModal').classList.add('open');
}

function closeCompareModal() {
  document.getElementById('compareModal').classList.remove('open');
}

async function runComparison() {
  var specIds = [];
  var submittalIds = [];
  document.querySelectorAll('#specCheckboxes input:checked').forEach(function(cb) { specIds.push(parseInt(cb.value)); });
  document.querySelectorAll('#submittalCheckboxes input:checked').forEach(function(cb) { submittalIds.push(parseInt(cb.value)); });

  if (!specIds.length) { showToast('Select at least one spec document'); return; }
  if (!submittalIds.length) { showToast('Select at least one submittal document'); return; }

  closeCompareModal();
  showToast('Running comparison... this may take a moment');

  try {
    var report = await apiAnalyzeDiscrepancies(activeProjectId, specIds, submittalIds);
    activeReport = report;
    await loadDiscrepancyReports();
    renderDiscrepancyReport(report);
    showDiscrepancyView();
    showToast('Comparison complete — ' + report.discrepancy_count + ' findings');
  } catch (e) {
    showToast('Comparison failed: ' + e.message);
  }
}

// ═══ REPORT VIEW ═══

async function viewDiscrepancyReport(reportId) {
  try {
    var report = await apiGetDiscrepancyReport(reportId);
    activeReport = report;
    renderDiscrepancyReport(report);
    showDiscrepancyView();
  } catch (e) {
    showToast('Failed to load report: ' + e.message);
  }
}

function showDiscrepancyView() {
  document.getElementById('chatView').classList.add('vh');
  document.getElementById('dashView').classList.add('vh');
  document.getElementById('discrepancyView').classList.remove('vh');
  document.getElementById('discrepancyView').style.display = '';
  document.querySelector('.input-area').style.display = 'none';
}

function renderDiscrepancyReport(report) {
  var view = document.getElementById('discrepancyView');

  var critCount = 0, majCount = 0, minCount = 0, infoCount = 0;
  (report.items || []).forEach(function(item) {
    if (item.severity === 'critical') critCount++;
    else if (item.severity === 'major') majCount++;
    else if (item.severity === 'minor') minCount++;
    else infoCount++;
  });

  var html = '<div class="disc-report">' +
    '<div class="disc-report-header">' +
    '<button class="disc-back" onclick="showChat();loadDiscrepancyReports()">← Back to chat</button>' +
    '<h2>' + esc(report.title) + '</h2>' +
    '<div class="disc-status disc-status-' + report.status + '">' + report.status + '</div>' +
    '</div>';

  if (report.summary) {
    html += '<div class="disc-summary">' + esc(report.summary) + '</div>';
  }

  html += '<div class="disc-stats">' +
    '<div class="disc-stat disc-stat-critical"><span class="disc-stat-num">' + critCount + '</span><span>Critical</span></div>' +
    '<div class="disc-stat disc-stat-major"><span class="disc-stat-num">' + majCount + '</span><span>Major</span></div>' +
    '<div class="disc-stat disc-stat-minor"><span class="disc-stat-num">' + minCount + '</span><span>Minor</span></div>' +
    '<div class="disc-stat disc-stat-info"><span class="disc-stat-num">' + infoCount + '</span><span>Info</span></div>' +
    '</div>';

  if (report.items && report.items.length) {
    html += '<div class="disc-items">';
    report.items.forEach(function(item) {
      html += renderDiscrepancyItem(item);
    });
    html += '</div>';
  } else {
    html += '<div style="padding:2rem;text-align:center;color:var(--text-dim)">No discrepancies found — documents appear to be in compliance.</div>';
  }

  html += '</div>';
  view.innerHTML = html;
}

function renderDiscrepancyItem(item) {
  var sevClass = 'disc-sev-' + item.severity;
  var sevLabel = item.severity.charAt(0).toUpperCase() + item.severity.slice(1);
  var catLabel = item.category.replace(/_/g, ' ');

  var specCite = item.spec_reference ? '<span class="doc-cite" data-doc-id="' + item.spec_doc_id + '" data-page="' + (item.spec_page || 1) + '">' + esc(item.spec_reference) + '</span>' : '';
  var submitCite = item.submittal_reference ? '<span class="doc-cite" data-doc-id="' + item.submittal_doc_id + '" data-page="' + (item.submittal_page || 1) + '">' + esc(item.submittal_reference) + '</span>' : '';

  var statusBtns = '';
  if (item.status === 'open') {
    statusBtns = '<button class="disc-action-btn" onclick="updateItemStatus(' + item.id + ',\'resolved\')">Resolve</button>' +
      '<button class="disc-action-btn disc-action-dismiss" onclick="updateItemStatus(' + item.id + ',\'dismissed\')">Dismiss</button>';
  } else {
    statusBtns = '<span class="disc-item-resolved">' + item.status + '</span>';
  }

  return '<div class="disc-item ' + sevClass + ' disc-item-' + item.status + '">' +
    '<div class="disc-item-header">' +
    '<span class="disc-sev-badge ' + sevClass + '">' + sevLabel + '</span>' +
    '<span class="disc-cat">' + catLabel + '</span>' +
    '<span class="disc-item-title">' + esc(item.title) + '</span>' +
    '<div class="disc-item-actions">' + statusBtns + '</div>' +
    '</div>' +
    '<div class="disc-item-body">' +
    '<p>' + esc(item.description) + '</p>' +
    (item.spec_excerpt || item.submittal_excerpt ? '<div class="disc-excerpts">' +
      (item.spec_excerpt ? '<div class="disc-excerpt"><div class="disc-excerpt-label">Spec says: ' + specCite + '</div><div class="disc-excerpt-text">' + esc(item.spec_excerpt) + '</div></div>' : '') +
      (item.submittal_excerpt ? '<div class="disc-excerpt disc-excerpt-sub"><div class="disc-excerpt-label">Submittal says: ' + submitCite + '</div><div class="disc-excerpt-text">' + esc(item.submittal_excerpt) + '</div></div>' : '') +
    '</div>' : '') +
    (item.recommendation ? '<div class="disc-recommendation"><strong>Recommendation:</strong> ' + esc(item.recommendation) + '</div>' : '') +
    '</div></div>';
}

async function updateItemStatus(itemId, newStatus) {
  try {
    await apiUpdateDiscrepancyItem(itemId, newStatus);
    if (activeReport) {
      var report = await apiGetDiscrepancyReport(activeReport.id);
      activeReport = report;
      renderDiscrepancyReport(report);
    }
    showToast('Item ' + newStatus);
  } catch (e) {
    showToast('Failed to update: ' + e.message);
  }
}
